### PR TITLE
8322166: Files.isReadable/isWritable/isExecutable expensive when file does not exist

### PR DESCRIPTION
**Pull request by bplb**
Add `isReadable`, `isWritable`, and `isExecutable` methods to `AbstractFileSystemProvider` and overrides to `UnixFileSystemProvider` and use these in the respective `Files` methods when the provider is an `AbstractFileSystemProvider`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322166](https://bugs.openjdk.org/browse/JDK-8322166): Files.isReadable/isWritable/isExecutable expensive when file does not exist (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [dc79cc2b](https://git.openjdk.org/jdk/pull/17133/files/dc79cc2b2acbf3214646179da30c7f871d4855d0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17133/head:pull/17133` \
`$ git checkout pull/17133`

Update a local copy of the PR: \
`$ git checkout pull/17133` \
`$ git pull https://git.openjdk.org/jdk.git pull/17133/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17133`

View PR using the GUI difftool: \
`$ git pr show -t 17133`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17133.diff">https://git.openjdk.org/jdk/pull/17133.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17133#issuecomment-1858584464)

### CODE REVIEW

**Comment by bplb** [comment]
The `access` method is modified to return the value of `errno` instead of relying on throwing an exception to denote success or failure. The `exists` method is changed to use `access` and the `exists0` native method is removed. The Windows implementation is not affected by the initial commit.

File: src/java.base/share/classes/java/nio/file/Files.java [hunk 1]
```diff
@@ -76,6 +76,7 @@
 import java.util.function.BiPredicate;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
+import sun.nio.fs.AbstractFileSystemProvider;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/file/Files.java:79]: Minor nit, you can probably group this with the imports of the other sun.nio classes.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java [hunk 2]
```diff
@@ -893,7 +893,9 @@ void move(UnixPath source, UnixPath target, CopyOption... options)
             sourceAttrs = UnixFileAttributes.get(source, false);
             if (sourceAttrs.isDirectory()) {
                 // ensure source can be moved
-                access(source, W_OK);
+                int errno = access(source, W_OK);
+                if (errno != 0)
+                    throw new UnixException(errno);
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java:898]: I don't think it's necessary to throw the UnixException here, should be able to use new UnixException(errno).rethrowAsIOException(source) to have it only throw the appropriate IOException for the error.

File: src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c [hunk 3]
```diff
@@ -1190,25 +1190,16 @@ Java_sun_nio_fs_UnixNativeDispatcher_realpath0(JNIEnv* env, jclass this,
...
     const char* path = (const char*)jlong_to_ptr(pathAddress);
 
     RESTARTABLE(access(path, (int)amode), err);
-    if (err == -1) {
-        throwUnixException(env, errno);
-    }
-}
 
-JNIEXPORT jboolean JNICALL
-Java_sun_nio_fs_UnixNativeDispatcher_exists0(JNIEnv* env, jclass this, jlong pathAddress) {
```
Comment (AlanBateman) [inline src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c:1207]: Shouldn't the UnixNativeDispatcher.exist method be removed too?

File: src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c [hunk 4]
```diff
@@ -1190,25 +1190,16 @@ Java_sun_nio_fs_UnixNativeDispatcher_realpath0(JNIEnv* env, jclass this,
...
-    int err;
-    const char* path = (const char*)jlong_to_ptr(pathAddress);
-    RESTARTABLE(access(path, F_OK), err);
-    return (err == 0) ? JNI_TRUE : JNI_FALSE;
+    return err == -1 ? errno : 0;
```
Comment (AlanBateman) [inline src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c:1202]: Style wise I think keep the parentheses to make it a bit clearer to read.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java [hunk 5]
```diff
@@ -350,12 +350,40 @@ public void checkAccess(Path obj, AccessMode... modes) throws IOException {
             mode |= X_OK;
         }
         try {
-            access(file, mode);
+            int errno = access(file, mode);
+            if (errno != 0)
+                throw new UnixException(errno);
         } catch (UnixException exc) {
             exc.rethrowAsIOException(file);
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:357]: This is another case where the UnixException doesn't need to be thrown.

**Review by AlanBateman — COMMENTED** [review]
This looks quite good and consistent with the how the "exists" method was implemented before FileSystemProvider was extended.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java [hunk 6]
```diff
@@ -350,12 +350,40 @@ public void checkAccess(Path obj, AccessMode... modes) throws IOException {
...
         }
     }
 
+    @Override
+    public boolean isReadable(Path path) {
+        UnixPath file = UnixPath.toUnixPath(path);
+        file.checkRead();
+        return access(file, R_OK) == 0 ? true : false;
```
Comment (liach) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:365]: Just curious, why do we use redundant `? true : false` for a boolean value? Such a pattern did not exist in this file before this patch.

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:361]: > why do we use redundant `? true : false` for a boolean value?

@liach What meaning do you intend by "redundant" here?

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java [hunk 7]
```diff
@@ -1027,7 +1029,10 @@ void copy(final UnixPath source,
             try {
                 // the access(2) system call always follows links so it
                 // is suppressed if the source is an unfollowed link
-                access(source, R_OK);
+                int errno = access(source, R_OK);
+                if (errno != 0)
+                    throw new UnixException(errno);
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java:1034]: Are you going to change this one to use `new UnixException(errno).rethrowAsIOException(source);` too?

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java:1034]: Yes. It was overlooked. Thanks.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java (hunk 6 above)
Comment (liach) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:361]: I mean we can just use `access(file, R_OK) == 0`

Comment (AlanBateman) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:361]: One of the iterations did change the "exists" methods, the three isXXXX methods were missed. It's not a correctness issue of course, just a cleanup.

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:361]: Thanks. I intentionally did not integrate it yet so I will fix those first.

**Pull request closed without merging**